docs/library/machine.CAN.rst

```diff
@@ -0,0 +1,364 @@
+errors.
+
+.. note:: The forthcoming ``can`` and ``aiocan`` micropython-lib modules are the
+          recommended way to use CAN with MicroPython.
+
+Constructor
+-----------
+
+.. class:: CAN(id, **kwargs)
+
+   Construct a CAN controller object of the given id:
+
+    - ``id`` identifies a particular CAN controller object; it is board and port specific.
+    - ``**kwargs`` are hardware-specific identifiers for the particular
+      board and port. For example, to set which GPIO pins to use for the
+      CAN controller.
+
+Methods
+-------
+
+.. method:: CAN.init(bitrate, mode=CAN.Mode.NORMAL, sample_point=75, sjw=None, tseg1=None, tseg2=None, f_clock=None)
+
+   Initialise the CAN bus with the given parameters:
+
+   - ``bitrate`` is the desired bus bit rate in bits per second.
+   - ``mode`` is one of :class:`CAN.Mode` enumerated values, indicating the
+     desired mode of operation.
+   - ``sample_point`` is a percentage of the data bit time. It specifies the
+     position of the bit sample with respect to the whole nominal bit time.
+
+   The remaining parameters all relate to CAN bit timings. If not specified, the CAN controller
+   will pick nominal values based on the specified bit rate.
+
+   Otherwise, they can be specified:
+
+    - ``sjw`` is the resynchronisation jump width in units of time quanta for nominal bits;
+      it can be a value between 1 and 4 inclusive for classic CAN, and between 1 and 128 inclusive for CAN FD.
+    - ``tseg1`` defines the location of the sample point in units of the time quanta for nominal bits;
+      it can be a value between 1 and 16 inclusive for classic CAN, and between 2 and 256 inclusive for CAN FD.
+    - ``tseg2`` defines the location of the transmit point in units of the time quanta for nominal bits;
+      it can be a value between 1 and 8 inclusive for classic CAN, and between 2 and 128 inclusive for CAN FD.
+    - ``f_clock`` is the system clock frequency for the CAN peripheral. This
+      can be omitted on many ports where the value is known to the hardware.
+
+    In this case, the CAN controller is configured correctly for the desired
+    ``bitrate`` and the specified total number of time quanta per bit.
+
+.. method:: CAN.init_fd(bitrate, sample_point=None, sjw=None, tseg1=None, tseg2=None, f_clock=None)
+
+   Initialise CAN FD for controllers which support it.
+
+   .. note:: This method is not present on the class for controllers that only
+             support CAN Classic.
+
+   This must be called after :func:`CAN.init()`. The parameters all have the
+   same meanings and behaviour as that function, however they configure the
+   Bit Rate Switch (BRS) feature of CAN FD.
+
+   When specifying exact timing parameters, the accepted ranges are:
+
+   - ``sjw`` can be a value between 1 and 16 inclusive.
+   - ``tseg1`` can be a value between 1 and 32 inclusive
+   - ``tseg2`` can be a value between 1 and 32 inclusive
+
+.. method:: CAN.detect_bitrate([timeout_ms])
```

Comment:
Unlikely to be implemented by most ports.

Reply:
This can easily be implemented by all ports. I have code I have written that generated the registers used to set bitrates for the majority of the CAN IC's made. it's a mater of iterating over the most common CAN speeds used collecting the registers for the chip that is being used. setting the registers in place. When messages start coming in and there are no errors you have located the correct bitrate. 

You can only do this for the typical bitrates because of how long it would take to run. If the user is supplying a timeout that number can be used to determine what bitrates are going to be tested.

Reply:
You don't typically want to cause errors on the CAN bus when turning on... so, you'd need to init the controller very carefully is listen only mode. I guess it's not hard... but, most HAL code from like ST or NXP doesn't allow you to do this easily. You'll have to work from the register set. That said, I guess I had to do the same for the NXP MIMXRT port to implement some of the requested features.

Reply:
I'll remove this. I'm not sure what my motivation for adding it was, I think I saw it in another driver's API and thought "oh, some controllers must support this as a hardware feature". But I they actually don't, and now I can't even find the driver where I saw it(!)

It should be possible to for someone to still add a higher-level function that works as @kdschlosser describes, without it being in the low-level driver API.

Reply:
Removed from the base classes.